### ISSUE TITLE
LookupError: backend "ibmqx4" is not available

### ISSUE THREAD

**Issue by anandumadhusoodanan** [QISKit/qiskit-sdk-py#449]
Whenever I try to run a program on the real device backend (in this case 'ibmqx4'), I get the above error.I have configured my API token using Qconfig as per the instructions given. I have ran the program on the local simulator and it runs fine. But whenever I run it on real device backend this shows up.
This happens even if the real device backends are available, which I verified by running the code as per the instructions and the results obtained are shown below

> [{'basisGates': 'u1,u2,u3,cx,id',
  'couplingMap': 'all-to-all',
  'id': '084e8de73c4d16330550c34cf97de3f2',
  'nQubits': 32,
  'name': 'ibmqx_hpc_qasm_simulator',
  'onlineDate': '2017-12-09T12:00:00.000Z',
  'serialNumber': 'hpc-simulator',
  'simulator': True,
  'status': 'on',
  'topologyId': '7ca1eda6c4bff274c38d1fe66c449dff'},
 {'basisGates': 'u1,u2,u3,cx,id',
  'chipName': 'Raven',
  'couplingMap': [[1, 0], [2, 0], [2, 1], [3, 2], [3, 4], [4, 2]],
  'description': '5 qubit transmon bowtie chip 3',
  'gateSet': 'SU2+CNOT',
  'id': 'c16c5ddebbf8922a7e2a0f5a89cac478',
  'nQubits': 5,
  'name': 'ibmqx4',
  'onlineDate': '2017-09-18T00:00:00.000Z',
  'serialNumber': 'ibmqx4',
  'simulator': False,
  'status': 'on',
  'topologyId': '3b8e671a5a3b56899e6e601e6a3816a1',
  'url': 'https://ibm.biz/qiskit-ibmqx4',
  'version': '1'},
 {'basisGates': 'u1,u2,u3,cx,id',
  'chipName': 'Albatross',
  'couplingMap': [[1, 0],
                  [1, 2],
                  [2, 3],
                  [3, 4],
                  [3, 14],
                  [5, 4],
                  [6, 5],
                  [6, 7],
                  [6, 11],
                  [7, 10],
                  [8, 7],
                  [9, 8],
                  [9, 10],
                  [11, 10],
                  [12, 5],
                  [12, 11],
                  [12, 13],
                  [13, 4],
                  [13, 14],
                  [15, 0],
                  [15, 2],
                  [15, 14]],
  'description': '16 transmon 2x8 ladder',
  'gateSet': 'SU2+CNOT',
  'id': 'f451527ae7b9c9998e7addf1067c0df4',
  'nQubits': 16,
  'name': 'ibmqx5',
  'onlineDate': '2017-09-21T00:00:00.000Z',
  'serialNumber': 'ibmqx5',
  'simulator': False,
  'status': 'on',
  'topologyId': 'ad8b182a0653f51dfbd5d66c33fd08c7',
  'url': 'https://ibm.biz/qiskit-ibmqx5',
  'version': '1'},
 {'basisGates': 'u1,u2,u3,cx,id',
  'chipName': 'Sparrow',
  'couplingMap': [[0, 1], [0, 2], [1, 2], [3, 2], [3, 4], [4, 2]],
  'description': '5 qubit transmon bowtie chip 2',
  'gateSet': 'SU2+CNOT',
  'id': '28147a578bdc88ec8087af46ede526e1',
  'nQubits': 5,
  'name': 'ibmqx2',
  'onlineDate': '2017-01-10T00:00:00.000Z',
  'serialNumber': 'Real5Qv2',
  'simulator': False,
  'status': 'on',
  'topologyId': '250e969c6b9e68aa2a045ffbceb3ac33',
  'url': 'https://ibm.biz/qiskit-ibmqx2',
  'version': '1'},
 {'basisGates': 'u1,u2,u3,cx,id',
  'couplingMap': 'all-to-all',
  'description': 'online qasm simulator',
  'id': '18da019106bf6b5a55e0ef932763a670',
  'nQubits': 24,
  'name': 'ibmqx_qasm_simulator',
  'simulator': True,
  'status': 'on',
  'topologyId': '250e969c6b9e68aa2a045ffbceb3ac33'}]

> 
It would be helpfull if I get to use the real device backend since it helps me with my post gradution project. Simulators only gets me ideal results.
I have used qiskit on python3 for running my code.

**Comment by ajavadia** [comment]
@anandumadhusoodanan can you tell us your version of qiskit (is it via pip install?) and a code snippet that you run to get the error?

**Comment by anandumadhusoodanan** [comment]
The version of qiskit is 0.4.11 ( installed via pip3)

The code that shows the error is 
```
import sys
try:
    sys.path.append("../../") 
    import Qconfig
    qx_config = {
        "APItoken": Qconfig.APItoken,
        "url": Qconfig.config['url']}
except:
    qx_config = {
        "APItoken":"4a5bd2cdfdc8e40d036aca804f68668e50e1b65fabed1579b182d63a1d8feb5ba911205b90fb956a1d4d7f5529913913eb98475e06d6ddf98accee666fe8d171",
        "url":"https://quantumexperience.ng.bluemix.net/api"}
    

from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit import QuantumProgram

from qiskit.tools.visualization import plot_histogram

from IBMQuantumExperience import IBMQuantumExperience
api = IBMQuantumExperience(token=qx_config['APItoken'], config={'url': qx_config['url']})


backend = 'ibmqx4' 
shots = 1024  

Q_program = QuantumProgram()

q2 = QuantumRegister('q2', 2)
c2 = ClassicalRegister('c2', 2)

bell = QuantumCircuit(q2, c2)
bell.h(q2[0])
bell.cx(q2[0], q2[1])

measureZZ = QuantumCircuit(q2, c2)
measureZZ.measure(q2[0], c2[0])
measureZZ.measure(q2[1], c2[1])

Q_program.add_circuit("bell_measureZZ", bell+measureZZ )

circuits = ["bell_measureZZ"]
result = Q_program.execute(circuits, backend=backend, shots=shots, max_credits=3, wait=10, timeout=240)
plot_histogram(result.get_counts("bell_measureZZ"))
result.get_data("bell_measureZZ")
```

**Comment by diego-plan9** [comment]
Can you append these lines to your sample:

```
from qiskit.backends import discover_remote_backends
backends = discover_remote_backends(api)
```

when not using `QuantumProgram.set_api()` (which is a good thing, as `QuantumProgram.set_api() is going to be deprecated soon), you need to explicitly discover the remote backends that are available before being able to use them. Please note that, in the next release, we are going to streamline the process further - please stay tuned when the new release is up if you plan to update `qiskit` :)

Also, can you regenerate your API token in https://quantumexperience.ng.bluemix.net/qx/experience, for security? The API Token should be kept secret and it seems it was present in your previous comment (I manually edited it) - regenerating and creating a new one will invalidate the old one.

**Comment by anandumadhusoodanan** [comment]
@diego-plan9 
I have regenerated the API (thanks for your information about that). 
I have also appended my sample with the above lines, but the following error shows up

> ResultError: QASM_NOT_VALID: Error parsing QASM. Error parsing qasm number 0. Identifiers must begin with a lowercase letter

**Comment by diego-plan9** [comment]
No worries - can you upgrade qiskit via `pip install -U qiskit`?  0.4.11 is lagging a bit behind (the latest version is 0.4.14), and the error you mention seems related to a fix introduced in 0.4.14.

**Comment by anandumadhusoodanan** [comment]
Yeah...The error got fixed !!
Thanks a lot for your help  @diego-plan9 @ajavadia
